File: filelist/src/main/scala/farjs/filelist/popups/SelectPopup.scala
package farjs.filelist.popups

import farjs.filelist.history.HistoryProvider
import farjs.filelist.popups.SelectController.selectPatternsHistoryKind
import farjs.ui._
import farjs.ui.popup.ModalContent._
import farjs.ui.popup._
import farjs.ui.theme.Theme
import scommons.react._
import scommons.react.hooks._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.scalajs.js

case class SelectPopupProps(showSelect: Boolean,
                            onAction: String => Unit,
                            onCancel: () => Unit)

object SelectPopup extends FunctionComponent[SelectPopupProps] {

  private[popups] var modalComp: ReactClass = Modal
  private[popups] var comboBoxComp: ReactClass = ComboBox

  protected def render(compProps: Props): ReactElement = {
    val historyProvider = HistoryProvider.useHistoryProvider
    val (maybeItems, setItems) = useState(Option.empty[js.Array[String]])
    val props = compProps.wrapped
    val (pattern, setPattern) = useState("")
    val (width, height) = (55, 5)
    val contentWidth = width - (paddingHorizontal + 2) * 2
    val contentLeft = 2
    val theme = Theme.useTheme().popup.regular

    val onAction: js.Function0[Unit] = { () =>
      if (pattern.nonEmpty) {
        props.onAction(pattern)
      }
    }
    
    useLayoutEffect({ () =>
      for {
        selectPatternsHistory <- historyProvider.get(selectPatternsHistoryKind).toFuture
        items <- selectPatternsHistory.getAll.toFuture
      } yield {
        val itemsReversed = items.reverse.map(_.item)
        itemsReversed.headOption.foreach { last =>
          setPattern(last)
        }
        setItems(Some(itemsReversed))
      }
      ()
    }, Nil)

    maybeItems.map { items =>
      <(modalComp)(^.plain := ModalProps(
        title =
          if (props.showSelect) "Select"
          else "Deselect",
        width = width,
        height = height,
        style = theme,
        onCancel = props.onCancel
      ))(
        <(comboBoxComp)(^.plain := ComboBoxProps(
          left = contentLeft,
          top = 1,
          width = contentWidth,
          items = items,
          value = pattern,
          onChange = { value =>
            setPattern(value)
          },
          onEnter = onAction
        ))()
      )
    }.orNull
  }
}

File: filelist/popups/SelectPopup.mjs
/**
 * @typedef {import("@farjs/ui/theme/Theme.mjs").ThemeEffects} ThemeEffects
 * @typedef {import("@farjs/filelist/history/HistoryProvider.mjs").HistoryKind} HistoryKind
 */
import React, { useLayoutEffect, useState } from "react";
import Modal from "@farjs/ui/popup/Modal.mjs";
import ModalContent from "@farjs/ui/popup/ModalContent.mjs";
import ComboBox from "@farjs/ui/ComboBox.mjs";
import Theme from "@farjs/ui/theme/Theme.mjs";
import HistoryProvider from "@farjs/filelist/history/HistoryProvider.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  readonly showSelect: boolean;
 *  onAction(pattern: string): void;
 *  onCancel(): void;
 * }} SelectPopupProps
 */

/**
 * @param {SelectPopupProps} props
 */
const SelectPopup = (props) => {
  const { modalComp, comboBoxComp } = SelectPopup;

  const historyProvider = HistoryProvider.useHistoryProvider();
  const [maybeItems, setItems] = useState(
    /** @type {string[] | undefined} */ (undefined)
  );
  const [pattern, setPattern] = useState("");
  const width = 55;
  const height = 5;
  const contentWidth = width - (ModalContent.paddingHorizontal + 2) * 2;
  const contentLeft = 2;
  /** @type {ThemeEffects} */
  const theme = Theme.useTheme().popup.regular;

  const onAction = () => {
    if (pattern.length > 0) {
      props.onAction(pattern);
    }
  };

  const readHistory = async () => {
    const selectPatternsHistory = await historyProvider.get(
      SelectPopup.selectPatternsHistoryKind
    );
    const items = await selectPatternsHistory.getAll();
    const itemsReversed = [...items].reverse().map((_) => _.item);
    if (itemsReversed.length > 0) {
      setPattern(itemsReversed[0]);
    }
    setItems(itemsReversed);
  };

  useLayoutEffect(() => {
    readHistory();
  }, []);

  if (maybeItems !== undefined) {
    return h(
      modalComp,
      {
        title: props.showSelect ? "Select" : "Deselect",
        width,
        height,
        style: theme,
        onCancel: props.onCancel,
      },
      h(comboBoxComp, {
        left: contentLeft,
        top: 1,
        width: contentWidth,
        items: maybeItems,
        value: pattern,
        onChange: (value) => {
          setPattern(value);
        },
        onEnter: onAction,
      })
    );
  }

  return null;
};

SelectPopup.displayName = "SelectPopup";
SelectPopup.modalComp = Modal;
SelectPopup.comboBoxComp = ComboBox;

/** @type {HistoryKind} */
SelectPopup.selectPatternsHistoryKind = {
  name: "farjs.selectPatterns",
  maxItemsCount: 50,
};

export default SelectPopup;
